drivers/clock_control/clock_control_mcux_syscon.c

```diff
@@ -187,15 +187,64 @@ static int mcux_lpc_syscon_clock_control_get_subsys_rate(
 		*rate = CLOCK_GetDmicClkFreq();
 		break;
 #endif
+#if defined(CONFIG_MEMC_MCUX_FLEXSPI)
+	case MCUX_FLEXSPI_CLK:
+#if (FSL_FEATURE_SOC_FLEXSPI_COUNT == 1)
+		*rate = CLOCK_GetFlexspiClkFreq();
+#else
+		*rate = CLOCK_GetFlexspiClkFreq(0);
+#endif
+		break;
+#if (FSL_FEATURE_SOC_FLEXSPI_COUNT == 2)
+	case MCUX_FLEXSPI2_CLK:
+		*rate = CLOCK_GetFlexspiClkFreq(1);
+		break;
+#endif
+#endif /* CONFIG_MEMC_MCUX_FLEXSPI */
 	}
 
 	return 0;
 }
 
+/*
+ * Since this function is used to reclock the FlexSPI when running in
+ * XIP, it must be located in RAM when MEMC driver is enabled.
+ */
+#ifdef CONFIG_MEMC
+#define SYSCON_SET_FUNC_ATTR __ramfunc
+#else
+#define SYSCON_SET_FUNC_ATTR
```

Comment:
This function is not adjusting the clock. It is handled inside `flexspi_clock_set_freq`. I think we don't need `mcux_lpc_syscon_clock_control_set_subsys_rate` to be in RAM.

Reply:
So the reason this function is in ram is that upon return from `flexspi_clock_set_freq`, the flash may not be accessible. The call sequence to change the clock looks like so:

- call `memc_mcux_flexspi_update_clock`. This function runs from RAM
- call `clock_control_set_rate`. This function will perform a flash read to get the API pointer for the `mcux_lpc_syscon_clock_control_set_subsys_rate` function
- call `flexspi_clock_set_freq`. ***After return from this function, flash access may not be possible***. If the clock rate change was significant enough, we will  need to update the DLL value of the FlexSPI before we can perform flash access again.
- call `FLEXSPI_UpdateDLLValue` and `memc_flexspi_reset`
- wait for the FlexSPI bus to be idle. After this point, we can safely access flash again.

For this reason, I think it is best to keep the `flexspi_clock_set_freq` function in RAM. It is hard to catch the bugs that result from invalid FlexSPI reads, which is why I typically err on the side of caution while changing the FlexSPI clock.

Reply:
What about the clock control API functions functions? Should we move those to RAM as well?

Reply:
The `clock_control_xxx` functions? Those functions are inlined, they will be declared in the scope of the file calling them (so they will be in RAM in this case). However, the implementations are not in RAM- that is why we have to reconfigure the FlexSPI before calling `clock_control_get_rate`: https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/memc/memc_mcux_flexspi.c#L120